src/tests/ftest/util/server_utils.py

```diff
@@ -622,3 +626,181 @@ def get_environment_value(self, name):
+            # The regex failed to get the rank and state
+            raise ServerFailed(
+                "Error obtaining {} output: {}".format(self.dmg, data))
+        try:
+            states = list(set([data[rank]["state"] for rank in data]))
+        except KeyError:
+            raise ServerFailed(
+                "Unexpected result from {} - missing 'state' key: {}".format(
+                    self.dmg, data))
+        if len(states) > 1:
+            # Multiple states for different ranks detected
+            raise ServerFailed(
+                "Multiple system states ({}) detected:\n  {}".format(
+                    states, data))
+        return states[0]
+
+    def check_system_state(self, valid_states, max_checks=1):
+        """Check that the DAOS system state is one of the provided states.
+
+        Fail the test if the current state does not match one of the specified
+        valid states.  Optionally the state check can loop multiple times,
+        sleeping one second between checks, by increasing the number of maximum
+        checks.
+
+        Args:
+            valid_states (list): expected DAOS system states as a list of
+                lowercase strings
+            max_checks (int, optional): number of times to check the state.
+                Defaults to 1.
+
+        Raises:
+            ServerFailed: if there was an error detecting the server state or
+                the detected state did not match one of the valid states
+
+        Returns:
+            str: the matching valid detected state
+
+        """
+        checks = 0
+        daos_state = "????"
+        while daos_state not in valid_states and checks < max_checks:
+            if checks > 0:
+                time.sleep(1)
+            try:
+                daos_state = self.get_single_system_state().lower()
+            except ServerFailed as error:
+                raise error
+            checks += 1
+            self.log.info("System state check (%s): %s", checks, daos_state)
+        if daos_state not in valid_states:
+            raise ServerFailed(
+                "Error checking DAOS state, currently neither {} after "
+                "{} state check(s)!".format(valid_states, checks))
+        return daos_state
+
+    def system_start(self):
+        """Start the DAOS IO servers.
+
+        Raises:
+            ServerFailed: if there was an error starting the servers
+
+        """
+        self.log.info("Starting DAOS IO servers")
+        self.check_system_state(("stopped"))
+        self.dmg.system_start()
+        if self.dmg.result.exit_status != 0:
+            raise ServerFailed(
+                "Error starting DAOS:\n{}".format(self.dmg.result))
+
+    def system_stop(self):
+        """Stop the DAOS IO servers.
+
+        Raises:
+            ServerFailed: if there was an error stopping the servers
+
+        """
+        self.log.info("Stopping DAOS IO servers")
+        self.check_system_state(("started", "joined"))
```

Comment:
Here, check_system_state raises an error if the system is already stopped. Isn't it better to check if the system is already stopped and if so, do nothing so that callers don't have to worry about whether the system is stopped before calling this method?

Reply:
That's a good point, this is some of the code I merged from Phils PR, I believe the intent is to assume that the servers are in a started/joined state before stop is called and if the servers are not in those states it must mean something went wrong before the user calls `system_stop`. 

Maybe the solution to this is to include an argument that allows the user to specify what state the server is assumed to be in before failing it. This way, both concerns are covered.